linode/helper/objects.go

```diff
@@ -20,14 +22,37 @@ func S3ConnFromResourceData(d *schema.ResourceData) *s3.S3 {
 	accessKey := d.Get("access_key").(string)
 	secretKey := d.Get("secret_key").(string)
 	cluster := d.Get("cluster").(string)
+	endpoint := d.Get("endpoint").(string)
 
-	sess := session.New(&aws.Config{
+	if endpoint == "" {
+		// fall back for legacy safety
+		endpoint = fmt.Sprintf(LinodeObjectsEndpoint, cluster)
+	}
+
+	return s3.New(session.New(&aws.Config{
 		// This region is hardcoded strictly for preflight validation purposes.
 		Region:      aws.String("us-east-1"),
 		Credentials: credentials.NewStaticCredentials(accessKey, secretKey, ""),
-		Endpoint:    aws.String(fmt.Sprintf(LinodeObjectsEndpoint, cluster)),
-	})
-	return s3.New(sess)
+		Endpoint:    aws.String(endpoint),
+	}))
+}
+
+// ComputeS3Endpoint sets endpoint in data for the s3 connection via the cluster and bucket combination
+func ComputeS3Endpoint(ctx context.Context, d *schema.ResourceData, meta interface{}) error {
+	if d.Get("endpoint").(string) != "" {
+		return nil
+	}
+
+	cluster := d.Get("cluster").(string)
+	bucket := d.Get("bucket").(string)
+
+	linodeClient := meta.(*ProviderMeta).Client
+	b, err := linodeClient.GetObjectStorageBucket(ctx, cluster, bucket)
+	if err != nil {
+		return fmt.Errorf("failed to find the specified Linode ObjectStorageBucket: %s", err)
+	}
+
+	return d.Set("endpoint", strings.TrimPrefix(b.Hostname, fmt.Sprintf("%s.", bucket)))
```

Comment:
Just to make the code a bit more clear, I wonder if it'd make sense to return the endpoint from this function and set it in state in the `readResource` function.

Reply:
per a conversation this is being done and the utils funcs will be rewritten a bit to return endpoints and the data sets will be done in the callers